pkg/kgo/compression.go

```diff
@@ -12,9 +12,12 @@ import (
 	"github.com/klauspost/compress/s2"
 	"github.com/klauspost/compress/zstd"
 	"github.com/pierrec/lz4/v4"
+	"github.com/twmb/franz-go/pkg/kgo/internal/pool"
 )
 
-var byteBuffers = sync.Pool{New: func() any { return bytes.NewBuffer(make([]byte, 8<<10)) }}
+const maxPoolDecodedBufferSize = 32 * 1024 * 1024 // 32MB
+
+var byteBuffers = sync.Pool{New: func() any { return bytes.NewBuffer(make([]byte, maxPoolDecodedBufferSize)) }}
```

Comment:
The main motivation is to avoid intermediate allocations resulting from batch decompression, given that the final size of the uncompressed buffer is unknown beforehand.

In the original implementation, this buffer is initialized with a size of `8k`, so if the uncompressed batch size is a few megabytes, the amount of garbage generated can be considerable, especially when the pool is drained.

Reply:
Taking this profile as a reference (see screenshot), basically what we are looking for is to get rid of all the allocated memory originated in the `s2.Decode` method.

![Screenshot 2024-09-25 at 17 43 10](https://github.com/user-attachments/assets/aba3db2f-c95a-4129-9a1b-2d3ffdd05d2f)

Reply:
I'm not sure but I think such a huge jump in size would be very difficult to get into upstream because it creates a lot of allocated overhead for applications that see much smaller batch sizes.

How about instead of allocating a fixed size we'd allocate a factor of the compressed size and then put that into the pool for reuse?

Reply:
We can try that approach and see how it performs in one some of our dev environments. 👍

Reply:
_changed @ 6d8cb09691d75e8c743d47e0d342d51afd0a675f_

Reply:
UPDATE: I've observed that after applying the suggested change the `decompress` function is once again among the top functions that allocate the most memory.

![Screenshot 2024-09-30 at 15 34 24](https://github.com/user-attachments/assets/1eccc58c-a094-40f7-a544-e985646ae3e9)

However, on the other hand, by reducing the maximum buffer size in the pool from 32MB to 8MB and using a dedicated pool for decoding the compressed batch, I have observed that the same benefit is achieved, so I would opt for this option instead. 

![Screenshot 2024-09-30 at 15 37 30](https://github.com/user-attachments/assets/c4f57736-384a-460d-bc56-c5e680a3fade)

As a side note, I've attached a screenshot where it is clearly observed how the sawtooth pattern, that indicates how often the GC triggers, extends over time after applying the change.


![Screenshot 2024-09-30 at 15 47 04](https://github.com/user-attachments/assets/c7026262-1445-4032-9a2c-d7f5ce9df243)

Reply:
_updated @ 6c4664377540ad88478f868c61f6c6f0b61c8fbc_